### PR TITLE
Load More For What I've Been Doing on Profiles

### LINKED ISSUE #495: Add Show More to Show full areas Addon
**Describe the new addon**

Adds a Load More button (to the "Show full areas" when "SHOW SCROLL FOR WHAT I'VE BEEN DOING" is enabled) to load more data from the user. I didn't add any styling for the button (besides the margining) so that's why it looks so scratch 2.0-ish. Try with "Scratch 2.0 → 3.0" it looks way better.

**Note**

Obviously, we can change where we enable and disable this option, but I just thought, for now, we could just merge them. But lmk if you guys think it's better another way.

**Link to the new addon**

~See what I added [Here!](https://github.com/TheColaber/ScratchAddonsUpdate/tree/GrahamSH-LLK-Animated-Thumbnail/addons/full-signature) - It's **fully** implemented and it works great.~

See what I added [Here!](https://github.com/TheColaber/ScratchAddonsUpdate/tree/TheColaber-full-siggy/addons/full-signature) - It's **fully** implemented and it works great.

### PR DESCRIPTION
**Pull request by GrahamSH-LLK**
**Resolves**

Resolves #495 

**Changes**

**Reason for changes**

More info is awesome
**Tests**

Console
**Complete Credit To @TheColaber**

### CODE REVIEW

File: addons/full-signature/working.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+export default async function ({ addon, global, console }) {
+  document.querySelector(".activity-stream").insertAdjacentHTML(
```
Comment (WorldLanguages) [inline addons/full-signature/working.js:2]: Please use appendChild

File: addons/full-signature/working.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    `
+  <button class="load-more-wibd">Load More</button>
+  `
+  );
+  let dataLoaded = 6;
+  let loadMore = document.querySelector(".load-more-wibd");
+  loadMore.addEventListener("click", function () {
+    fetch(
+      `https://scratch.mit.edu/messages/ajax/user-activity/?user=${
+        document.querySelector(".header-text h2").innerHTML
```
Comment (WorldLanguages) [inline addons/full-signature/working.js:13]: You can use `textContent` here - or just look at the URL

File: addons/full-signature/working.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      }&max=1000000`
+    )
+      .then(function (response) {
+        return response.text();
+      })
+      .then(function (text) {
+        let dummyEl = document.createElement("div");
+        dummyEl.innerHTML = text;
+        let lastDataLoad = dataLoaded;
+        for (; dataLoaded < lastDataLoad + 6; dataLoaded++) {
```
Comment (WorldLanguages) [inline addons/full-signature/working.js:23]: lol I've never seen a for with a semicolon at the beginning, but I know it works

File: addons/full-signature/working.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+          loadMore.insertAdjacentHTML(
```
Comment (WorldLanguages) [inline addons/full-signature/working.js:24]: Use appendChild

File: addons/full-signature/working.js (hunk 3 above)
Comment (TheColaber) [reply addons/full-signature/working.js:23]: Without it, it gives an error

**Comment by TheColaber** [comment]
Graham, Do you want to do the changes? I think just change the insertAdjacentHTMs to appendChild. For the for loop I don't think you can remove the semicolon without getting an error.

**Comment by WorldLanguages** [comment]
@TheColaber I didn't ask for any changes on that line. It does break if you remove the semicolon

**Comment by TheColaber** [comment]
> @TheColaber I didn't ask for any changes on that line. It does break if you remove the semicolon

It showed up as a review so I assumed...

**Comment by GrahamSH-LLK** [comment]
@TheColaber Can you fix the above problems?

**Comment by TheColaber** [comment]
> @TheColaber Can you fix the above problems?

@GrahamSH-LLK How can I edit your branch? Would I have to make a copy of it?

File: addons/full-signature/working.js (hunk 1 above)
Comment (Hans5958) [reply addons/full-signature/working.js:2]: Nothing wrong with insertAdjacentHTML. It is also safe.

**Comment by Hans5958** [comment]
@TheColaber, you can create a pull request that points to `GrahamSH-LLK-lol`, and I can merge it quickly.

**Comment by TheColaber** [comment]
> @TheColaber, you can create a pull request that points to `GrahamSH-LLK-lol`, and I can merge it quickly.

Ok Creating pr now.

**Comment by TheColaber** [comment]
How can you insert plain text HTML without using insertAdjacentHTMs using appendChild() requires it to be a node.

**Comment by easrng** [comment]
> How can you insert plain text HTML without using insertAdjacentHTMs using appendChild() requires it to be a node.

`document.createTextNode("Hi!")`

**Comment by JeffaloBob** [comment]
> > How can you insert plain text HTML without using insertAdjacentHTMs using appendChild() requires it to be a node.
> 
> `document.createTextNode("Hi!")`

this

**Comment by TheColaber** [comment]
> > > How can you insert plain text HTML without using insertAdjacentHTMs using appendChild() requires it to be a node.
> > 
> > 
> > `document.createTextNode("Hi!")`
> 
> this

Never knew that existed, thx.

**Comment by Explosion-Scratch** [comment]
Is adding this spying on people? I mean... It sort of feels a little
'stalky'

**Comment by TheColaber** [comment]
I mean you can already just look at the API... this just makes it easier lol

**Comment by TheColaber** [comment]
> createTextNode
@JeffaloBob but it still shows it as text even though it is now a node.

**Comment by GrahamSH-LLK** [comment]
Createtextnode creates a _text_ node. It doesn't accept any HTML. Is that what you are asking?

**Comment by TheColaber** [comment]
> Createtextnode creates a _text_ node. It doesn't accept any HTML. Is that what you are asking?

oh lol... I didn't see _text_ I'm am taking the HTML from the API which returns a string (that is HTML)

For instance, it might return: 
```
"<!-- templates/notifications/includes/user-feed.html --> <ul class="activity-stream"> <li> <span class="icon-xs black love"></span> <div> <span class="actor">TheColaber</span> loved <a href="/projects/434107175/">Spriteless Tetris</a> <span data-tag="time" class="time">5 hours, 4 minutes ago</span> </div> </li> <li> <span class="icon-xs black follow"></span> <div> <span class="actor">TheColaber</span> was promoted to manager of <a href="/studios/27732033/">Scratch city | Collab @CristoferK and @TheColaber </a> <span data-tag="time" class="time">1 week, 3 days ago</span> </div> </li> <li> <span class="icon-xs black follow"></span> <div> <span class="actor">TheColaber</span> became a curator of <a href="/studios/27732033/">Scratch city | Collab @CristoferK and @TheColaber </a> <span data-tag="time" class="time">1 week, 3 days ago</span> </div> </li> <li> <span class="icon-xs black love"></span> <div> <span class="actor">TheColaber</span> loved <a href="/projects/432914869/">Basketball</a>..."
```

So I am trying to figure out how to turn this into a node so I can use `appendChild`

**Comment by GrahamSH-LLK** [comment]
If you have the HTML in a variable you can just $('<selector>').appendChild(variablewithhtml)

**Comment by TheColaber** [comment]
> $.get( "link", function( data ) {
>   $( "selector" ).append( data );
> })

Isn't that jQuery?

**Comment by GrahamSH-LLK** [comment]
It is, but the profile page already has jquery, so you should be good.

**Comment by TheColaber** [comment]
That is a workaround for fetching but that also still gets the data as a string.

**Comment by TheColaber** [comment]
Here is the code:
```javascript
  let activityStream = document.querySelectorAll(".activity-stream li")
  let username = document.querySelector(".header-text h2").textContent
  let loadMore = document.createElement("button")
  loadMore.classList.add("load-more-wibd")
  loadMore.innerText = "Load More"

  document.querySelector(".activity-stream").appendChild(loadMore);
  let dataLoaded = 6;
  loadMore.addEventListener("click", function() {
    let lastDataLoad = dataLoaded;
    for (; dataLoaded < lastDataLoad + 6; dataLoaded++) {
      $.get(`https://scratch.mit.edu/messages/ajax/user-activity/?user=${
            username}&max=1000000`, function(data) {
        activityStream[activityStream.length - 1].append(data);
      })
    }

  });
```

**Comment by GrahamSH-LLK** [comment]
Does that work?

**Comment by TheColaber** [comment]
> Does that work?

No, the text needs to be converted to HTML... the only solution I see is using a parser but we can't use it.

**Comment by TheColaber** [comment]
@WorldLanguages Is jQuery.parseHTML() allowed?

**Comment by WorldLanguages** [comment]
@TheColaber No. As a rule of thumb, if you need to write any of these symbols, it's not allowed: <>

**Comment by Hans5958** [comment]
> 
> 
> > Createtextnode creates a _text_ node. It doesn't accept any HTML. Is that what you are asking?
> 
> oh lol... I didn't see _text_. I'm am taking the HTML from the API which returns a string (that is HTML)
> 
> For instance, it might return:
> 
> ```
> " <ul class="activity-stream"> <li> <span class="icon-xs black love"></span> <div> <span class="actor">TheColaber</span> loved <a href="/projects/434107175/">Spriteless Tetris</a> <span data-tag="time" class="time">5 hours, 4 minutes ago</span> </div> </li> <li> <span class="icon-xs black follow"></span> <div> <span class="actor">TheColaber</span> was promoted to manager of <a href="/studios/27732033/">Scratch city | Collab @CristoferK and @TheColaber </a> <span data-tag="time" class="time">1 week, 3 days ago</span> </div> </li> <li> <span class="icon-xs black follow"></span> <div> <span class="actor">TheColaber</span> became a curator of <a href="/studios/27732033/">Scratch city | Collab @CristoferK and @TheColaber </a> <span data-tag="time" class="time">1 week, 3 days ago</span> </div> </li> <li> <span class="icon-xs black love"></span> <div> <span class="actor">TheColaber</span> loved <a href="/projects/432914869/">Basketball</a>..."
> ```
> 
> So I am trying to figure out how to turn this into a node so I can use `appendChild`

The only logical thing is to use `insertAdjacentHTML()`, but I have no clue why you guys think it is not safe. (spoiler: it is safe, innerHTML/outerHTML is not)

**Comment by Explosion-Scratch** [comment]
I guess so lol

**Comment by mxmou** [comment]
> The only logical thing is to use `insertAdjacentHTML()`, but I have no clue why you guys think it is not safe. (spoiler: it is safe, innerHTML/outerHTML is not)

Anything that can parse HTML is not safe.

**Comment by TheColaber** [comment]
> > The only logical thing is to use `insertAdjacentHTML()`, but I have no clue why you guys think it is not safe. (spoiler: it is safe, innerHTML/outerHTML is not)
> 
> Anything that can parse HTML is not safe.

What can I use to insert text like this "<a href='scratch.mit.edu'>Go to Scratch" on to a certain place on the screen.

**Comment by GrahamSH-LLK** [comment]
$('selector').append('<a href="https://scratch.mit.edu">Go To Scratch</a>') on a page with jquery, or document.querySelector('selector').appendChild('<a href="https://scratch.mit.edu">Go To Scratch</a>') on a page without.

**Comment by Hans5958** [comment]
> 
> 
> > The only logical thing is to use `insertAdjacentHTML()`, but I have no clue why you guys think it is not safe. (spoiler: it is safe, innerHTML/outerHTML is not)
> 
> Anything that can parse HTML is not safe.

Eh, if you wanted to see that way, then it depends. If the HTML string that will be inserted is static, it is obviously safe (as safe as innerHTML). It won't be safe if the HTML can be controlled by the user. However, this does not mean that easy mitigation like DOM purifying is not possible.

**Comment by TheColaber** [comment]
> > > The only logical thing is to use `insertAdjacentHTML()`, but I have no clue why you guys think it is not safe. (spoiler: it is safe, innerHTML/outerHTML is not)
> > 
> > 
> > Anything that can parse HTML is not safe.
> 
> What can I use to insert text like this "[Go to Scratch" on to a certain place on the screen.](scratch.mit.edu)



> > > The only logical thing is to use `insertAdjacentHTML()`, but I have no clue why you guys think it is not safe. (spoiler: it is safe, innerHTML/outerHTML is not)
> > 
> > 
> > Anything that can parse HTML is not safe.
> 
> What can I use to insert text like this "[Go to Scratch" on to a certain place on the screen.](scratch.mit.edu)



> $('selector').append('[Go To Scratch](https://scratch.mit.edu)') on a page with jquery, or document.querySelector('selector').appendChild('[Go To Scratch](https://scratch.mit.edu)') on a page without.

Whoops, I think github turned my HTML to text... Here's what I wanted to insert
```
<a href="scratch.mit.edu>Go to Scratch</a>
```
Not text

**Comment by GrahamSH-LLK** [comment]
In this case, I think we should be pretty safe because the values are sanitized by scratch.

**Comment by TheColaber** [comment]
> In this case, I think we should be pretty safe because the values are sanitized by scratch.

To use `insertAdjecentHTML()`?

**Comment by Hans5958** [comment]
Yeah.

----

On an unrelated note, wouldn't it be simpler if you do this instead?

```js
fetch(`...`)
	.then(response => response.text())
	.then(response => {
		var html = new DOMParser().parseFromString(response, 'text/html');
		html.querySelectorAll("ul > li").forEach((li, index) => {
			if (index > 5) document.querySelector(".activity-stream").appendChild(li)
		})
	})
```

**Comment by TheColaber** [comment]
> Yeah.
> 
> On an unrelated note, wouldn't it be simpler if you do this instead?
> 
> ```js
> fetch(`...`)
> 	.then(response => response.text())
> 	.then(response => {
> 		var html = new DOMParser().parseFromString(response, 'text/html');
> 		html.querySelectorAll("ul > li").forEach((li, index) => {
> 			if (index > 5) document.querySelector(".activity-stream").appendChild(li)
> 		})
> 	})
> ```

It would, however, I don't think they want me using a parser... Anyway, I'll use that code for now then.

**Comment by Hans5958** [comment]
Well, I have the vote for you. I think it is okay to use `DOMParser()` instead of `innerHTML`.

**Comment by GrahamSH-LLK** [comment]
In this case, it's actually quite safe.

**Comment by TheColaber** [comment]
K, this code seems to be working:
```javascript
let activityStream = document.querySelectorAll(".activity-stream li");
  let parser = new DOMParser();
  let htmlparse = parser.parseFromString(
    `
    <button class="load-more-wibd">Load More</button>
    `,
    "text/html"
  );
  document.querySelector(".activity-stream").appendChild(htmlparse.querySelector("button"));
  let dataLoaded = 6;
  let loadMore = document.querySelector(".load-more-wibd");
  loadMore.addEventListener("click", function() {
    fetch(`
      https://scratch.mit.edu/messages/ajax/user-activity/?user=${
      document.querySelector(".header-text h2").textContent
      }&max=1000000`)
      .then(response => response.text())
      .then(response => {
        var html = parser.parseFromString(response, 'text/html');
        html.querySelectorAll("ul > li").forEach((li, index) => {
          if (index > dataLoaded && index < dataLoaded + 6) {
            activityStream[activityStream.length - 1].appendChild(li)
          }
        })
        dataLoaded += 6
      })
  })
```
But now I just need to get @WorldLanguages to approve of this.

**Comment by Hans5958** [comment]
Just update the PR

**Comment by TheColaber** [comment]
> Just update the PR

Can you merge it once I create it?

**Comment by mxmou** [comment]
As @WorldLanguages wrote:

> The issue isn't innerHTML itself really (even tho it's dangerous if you don't know what you're doing), but the review process in extension stores. They don't like innerHTML.

If they don't like `innerHTML` they probably don't like `DOMParser` either.

**Comment by GrahamSH-LLK** [comment]
I did it for you because it's easier to just commit.

**Comment by TheColaber** [comment]
> As @WorldLanguages wrote:
> 
> > The issue isn't innerHTML itself really (even tho it's dangerous if you don't know what you're doing), but the review process in extension stores. They don't like innerHTML.
> 
> If they don't like `innerHTML` they probably don't like `DOMParser` either.

We have got rid of all `.innerHTML`s but I don't know what can replace the `DOMParser` without either using `.insertAdjectHTML()` or `.innerHTML = ""`

**Comment by Hans5958** [comment]
> 
> 
> As @WorldLanguages wrote:
> 
> > The issue isn't innerHTML itself really (even tho it's dangerous if you don't know what you're doing), but the review process in extension stores. They don't like innerHTML.
> 
> If they don't like `innerHTML` they probably don't like `DOMParser` either.

Not sure if that's the case, that is...

**Comment by WorldLanguages** [comment]
DOMParser is such a hacky way to avoid using `innerHTML`, tho. If you're not going to use `createElement` and `appendChild` I'd rather see you use non-hacky innerHTML

**Comment by Hans5958** [comment]
Looking at the context of the implementation, I have no reason why I should going to the hassle of creating a new element and trying to replicate the same element, when I can just copy the elements that they have prepared on the server for the exact use.

btw, we can do `innerHTML` or `insertAdjacentHTML`.

**Comment by Explosion-Scratch** [comment]
Maybe instead of load more, just infinite scrolling?

**Comment by Hans5958** [comment]
> Maybe instead of load more, just infinite scrolling?

That is @lholliger's addon.

**Comment by GrahamSH-LLK** [comment]
I do like infinite scrolling, though. We could just innerhtml it with the entire list!

**Comment by Explosion-Scratch** [comment]
Sure, but isn't innerHTML not allowed by extension stores?

**Comment by Hans5958** [comment]
> 
> Sure, but isn't innerHTML not allowed by extension stores?

If it is, then why is it still accepted, despite [Vue, and some addons, also has innerHTML](https://user-images.githubusercontent.com/11584103/97019521-9da59500-157a-11eb-93b7-c91747dcc147.png)?

I think it's more like it is more it will give a warn than an outright rejection.

**Comment by Explosion-Scratch** [comment]
Okay, I didn't know that.

**Review by ljy-002 — CHANGES_REQUESTED** [review]
.activity-stream {
  overflow-y: scroll !important;
}
ul.activity-stream div {
  width: 90%;
}

.load-more-wibd{
  margin-left:25%;
  margin-right:25%;
  color: blue
}

**Comment by ljy-002** [comment]
I think this should be more beautiful

**Comment by ljy-002** [comment]
[CSS](https://github.com/ScratchAddons/ScratchAddons/pull/499/commits/333a0a0813ca1049bf196eff7d6ba2a517850ebd)

**Comment by TheColaber** [comment]
Alright, I did a bit of work now we are down to just fixing line 15 of this code below for this to be allowed:
```javascript
export default async function ({ addon, global, console }) {
  let activityStream = document.querySelectorAll(".activity-stream li");
  let loadMore = document.createElement("button")
  loadMore.classList.add("load-more-wibd")
  loadMore.innerText = "Load More"
  document.querySelector(".activity-stream").appendChild(loadMore);
  let dataLoaded = 6;
  loadMore.addEventListener("click", function () {
    fetch(`
      https://scratch.mit.edu/messages/ajax/user-activity/?user=${
        document.querySelector(".header-text h2").textContent
      }&max=1000000`)
      .then((response) => response.text())
      .then((response) => {
        var html = new DOMParser().parseFromString(response, "text/html"); // This line must not have a DOMParser... what can we do??
        html.querySelectorAll("ul > li").forEach((li, index) => {
          if (index > dataLoaded && index < dataLoaded + 6) {
            activityStream[activityStream.length - 1].append(li);
          }
        });
        dataLoaded += 6;
      });
  });
}

```
Right now, I see no solution to line 15, as the HTML comes back as text and cannot use append or appendChild or createElement

**Comment by easrng** [comment]
> Alright, I did a bit of work now we are down to just fixing line 15 of this code below for this to be allowed:
> 
> ```js
> Right now, I see no solution to line 15, as the HTML comes back as text and cannot use append or appendChild or createElement

Take this:
```js
function wiwoDOMToObjects(d) {
    [...d.querySelectorAll("li")].map(e => {
        const actor = e.querySelector(".actor")
        const time = e.querySelector(".time")
        const o = {
            actor: actor.textContent.trim(),
            time: time.textContent.trim(),
            icon: [...e.querySelector("[class^=icon]").classList].filter(e=>!(e.startsWith("icon")||e=="black"))[0]
        }
        actor.remove()
        time.remove()
        o.text = [...e.querySelector("div").childNodes].filter(e => e.nodeType == document.TEXT_NODE || (e.nodeType == document.ELEMENT_NODE && e.tagName.toLowerCase() == "a")).map(e => ({
            text: e.textContent.replace(/\s+/g, " "),
            link: e.href || undefined
        })).filter(e => e.text.trim())
        return o
    })
}

function wiwoTextToEle(ele) {
    return text.map(text => {
        let el;
        if (e.link) {
            el = document.createElement("a")
            el.href = e.link
            el.textContent = e.text
        } else {
            el = document.createTextNode(e.text)
        }
        return el
    })
}

export default async function ({ addon, global, console }) {
  let activityStream = document.querySelectorAll(".activity-stream li");
  let loadMore = document.createElement("button")
  loadMore.classList.add("load-more-wibd")
  loadMore.innerText = "Load More"
  document.querySelector(".activity-stream").appendChild(loadMore);
  let dataLoaded = 6;
  loadMore.addEventListener("click", function () {
    fetch(`
      https://scratch.mit.edu/messages/ajax/user-activity/?user=${
        document.querySelector(".header-text h2").textContent
      }&max=1000000`)
      .then((response) => response.text())
      .then((response) => {
        var doc = new DOMParser().parseFromString(response, "text/html"); // This line must not have a DOMParser... what can we do??
        wiwoDOMToObjects(doc).forEach((o, index) => {
          if (index > dataLoaded && index < dataLoaded + 6) {
            // turn the object into a new HTML element
            activityStream[activityStream.length - 1].append(...wiwoTextToEle(o.text));
          }
        });
        dataLoaded += 6;
      });
  });
}
```

**Comment by TheColaber** [comment]
There are 2 problems with this script...

It still has a DOMParcer
I am getting an error that says 
```
Uncaught (in promise) TypeError: Cannot read property 'forEach' of undefined
    at working.js:49
```

**Comment by WorldLanguages** [comment]
You can use a DOMParser to parse HTML - just don't use it as a way to avoid using createElement

**Comment by TheColaber** [comment]
@GrahamSH-LLK would you like me to create a pull request?

**Comment by GrahamSH-LLK** [comment]
Oh, sorry. @TheColaber, you can PR or I can copy that code.

**Comment by TheColaber** [comment]
Copying the code would be easier...

**Comment by WorldLanguages** [comment]
This PR got too chaotic. Please open another one from "scratch" hehe :p

**Pull request closed without merging**